File: src/lib/transfer.ts
import fs from "fs";
import path from "path";
import { Client as SSHClient } from "ssh2";
import { db } from "./db";
import { jobs, groups, destinations, subtitleFiles, settings } from "./db/schema";
import { eq } from "drizzle-orm";
import { formatGroupedPath } from "./naming";
import type { Job, Destination } from "./db/schema";

const MAX_CONCURRENT = 2;
let activeTransfers = 0;
const transferQueue: Array<{ jobId: number; destinationId: number }> = [];

function updateJobProgress(
  jobId: number,
  progress: number,
  error?: string | null
) {
  const status =
    error ? "failed" : progress >= 1 ? "completed" : "transferring";
  db.update(jobs)
    .set({
      transferProgress: progress,
      transferError: error ?? null,
      status,
      updatedAt: new Date().toISOString(),
    })
    .where(eq(jobs.id, jobId))
    .run();
}

/**
 * Build the relative destination path for a job using group context and naming presets.
 */
function buildRelativePath(job: Job): string {
  // Look up the group for this job
  const group = job.groupId
    ? db.select().from(groups).where(eq(groups.id, job.groupId)).get()
    : null;

  // Look up naming settings
  const namingPreset =
    db.select().from(settings).where(eq(settings.key, "naming_preset")).get()?.value || "jellyfin";
  const specialsFolderName =
    db.select().from(settings).where(eq(settings.key, "specials_folder_name")).get()?.value || "Specials";
  const extrasFolderName =
    db.select().from(settings).where(eq(settings.key, "extras_folder_name")).get()?.value || "Extras";

  if (group) {
    return formatGroupedPath(job, group, {
      naming_preset: namingPreset,
      specials_folder_name: specialsFolderName,
      extras_folder_name: extrasFolderName,
    });
  }

  // Fallback for ungrouped jobs: construct a minimal group-like object
  return formatGroupedPath(job, {
    id: 0,
    status: "matched",
    mediaType: job.mediaType,
    folderPath: "",
    folderName: "",
    totalFileCount: 1,
    totalFileSize: job.fileSize,
    parsedTitle: job.parsedTitle,
    parsedYear: job.parsedYear,
    tmdbId: job.tmdbId,
    tmdbTitle: job.tmdbTitle,
    tmdbYear: job.tmdbYear,
    tmdbPosterPath: job.tmdbPosterPath,
    matchConfidence: job.matchConfidence,
    destinationId: job.destinationId,
    createdAt: job.createdAt,
    updatedAt: job.updatedAt,
  }, {
    naming_preset: namingPreset,
    specials_folder_name: specialsFolderName,
    extras_folder_name: extrasFolderName,
  });
}

/**
 * Build subtitle destination path by replacing the media extension with
 * the subtitle extension (and optional language code).
 * E.g. "Show S01E01 - Title.mkv" → "Show S01E01 - Title.en.srt"
 */
function buildSubtitlePath(
  mediaRelativePath: string,
  subtitleExt: string,
  languageCode: string | null
): string {
  const lastDot = mediaRelativePath.lastIndexOf(".");
  const basePath = lastDot !== -1 ? mediaRelativePath.slice(0, lastDot) : mediaRelativePath;
  const ext = subtitleExt.startsWith(".") ? subtitleExt : `.${subtitleExt}`;
  return languageCode ? `${basePath}.${languageCode}${ext}` : `${basePath}${ext}`;
}

/**
 * Transfer subtitle files for a completed job (local copy, fire-and-forget)
 */
function transferSubtitlesLocal(
  jobId: number,
  mediaRelativePath: string,
  destBasePath: string
) {
  const subs = db
    .select()
    .from(subtitleFiles)
    .where(eq(subtitleFiles.jobId, jobId))
    .all();

  for (const sub of subs) {
    try {
      const subRelPath = buildSubtitlePath(
        mediaRelativePath,
        sub.fileExtension,
        sub.languageCode
      );
      const fullDest = path.join(destBasePath, subRelPath);
      fs.mkdirSync(path.dirname(fullDest), { recursive: true });
      fs.copyFileSync(sub.sourcePath, fullDest);
    } catch (err) {
      console.error(
        `Failed to copy subtitle ${sub.sourcePath}:`,
        err instanceof Error ? err.message : err
      );
    }
  }
}

/**
 * Transfer subtitle files for a completed job via SFTP
 */
async function transferSubtitlesSFTP(
  jobId: number,
  mediaRelativePath: string,
  dest: Destination
): Promise<void> {
  const subs = db
    .select()
    .from(subtitleFiles)
    .where(eq(subtitleFiles.jobId, jobId))
    .all();

  if (subs.length === 0) return;

  return new Promise((resolve, reject) => {
    const conn = new SSHClient();

    conn.on("ready", () => {
      conn.sftp((err, sftp) => {
        if (err) {
          conn.end();
          reject(err);
          return;
        }

        let completed = 0;
        const total = subs.length;

        for (const sub of subs) {
          const subRelPath = buildSubtitlePath(
            mediaRelativePath,
            sub.fileExtension,
            sub.languageCode
          );
          const fullDest =
            dest.basePath.replace(/\\/g, "/") +
            "/" +
            subRelPath.replace(/\\/g, "/");

          // Create remote directories then upload
          const dirs = path.dirname(fullDest).split("/").filter(Boolean);
          let currentDir = "/";
          const mkdirAndUpload = (index: number) => {
            if (index >= dirs.length) {
              // Upload the file
              const readStream = fs.createReadStream(sub.sourcePath);
              const writeStream = sftp.createWriteStream(fullDest);
              writeStream.on("close", () => {
                completed++;
                if (completed === total) {
                  conn.end();
                  resolve();
                }
              });
              writeStream.on("error", (writeErr: Error) => {
                console.error(
                  `Failed to SFTP subtitle ${sub.sourcePath}:`,
                  writeErr.message
                );
                completed++;
                if (completed === total) {
                  conn.end();
                  resolve();
                }
              });
              readStream.pipe(writeStream);
              return;
            }
            currentDir += (currentDir === "/" ? "" : "/") + dirs[index];
            sftp.mkdir(currentDir, () => mkdirAndUpload(index + 1));
          };
          mkdirAndUpload(0);
        }
      });
    });

    conn.on("error", (connErr) => {
      console.error("SFTP subtitle connection error:", connErr.message);
      resolve(); // Don't fail the main transfer for subtitle errors
    });

    const connectConfig: Record<string, unknown> = {
      host: dest.sshHost!,
      port: dest.sshPort || 22,
      username: dest.sshUser!,
    };

    if (dest.sshKeyPath) {
      connectConfig.privateKey = fs.readFileSync(dest.sshKeyPath);
      if (dest.sshKeyPassphrase) {
        connectConfig.passphrase = dest.sshKeyPassphrase;
      }
    }

    conn.connect(connectConfig);
  });
}

/**
 * Local file copy with progress tracking
 */
async function transferLocal(
  job: Job,
  dest: Destination
): Promise<void> {
  const relativePath = buildRelativePath(job);
  const fullDest = path.join(dest.basePath, relativePath);

  // Create directory structure
  fs.mkdirSync(path.dirname(fullDest), { recursive: true });

  const totalSize = job.fileSize;
  let transferred = 0;

  // Check for partial file (resume support)
  if (fs.existsSync(fullDest)) {
    const existingStat = fs.statSync(fullDest);
    if (existingStat.size === totalSize) {
      // Already complete
      updateJobProgress(job.id, 1);
      return;
    }
    if (existingStat.size < totalSize) {
      transferred = existingStat.size;
    }
  }

  return new Promise((resolve, reject) => {
    const readStream = fs.createReadStream(job.sourcePath, {
      start: transferred,
    });
    const writeStream = fs.createWriteStream(fullDest, {
      flags: transferred > 0 ? "a" : "w",
    });

    readStream.on("data", (chunk) => {
      transferred += chunk.length;
      const progress = Math.min(transferred / totalSize, 1);
      updateJobProgress(job.id, progress);
    });

    readStream.on("error", (err) => {
      updateJobProgress(job.id, transferred / totalSize, err.message);
      reject(err);
    });

    writeStream.on("error", (err) => {
      updateJobProgress(job.id, transferred / totalSize, err.message);
      reject(err);
    });

    writeStream.on("finish", () => {
      updateJobProgress(job.id, 1);

      // Save destination path on the job
      db.update(jobs)
        .set({
          destinationId: dest.id,
          destinationPath: fullDest,
          updatedAt: new Date().toISOString(),
        })
        .where(eq(jobs.id, job.id))
        .run();

      resolve();
    });

    readStream.pipe(writeStream);
  });
}

/**
 * SFTP transfer with progress tracking
 */
async function transferSFTP(
  job: Job,
  dest: Destination
): Promise<void> {
  const relativePath = buildRelativePath(job);
  // Use forward slashes for remote path
  const fullDest = dest.basePath.replace(/\\/g, "/") + "/" + relativePath.replace(/\\/g, "/");

  return new Promise((resolve, reject) => {
    const conn = new SSHClient();

    conn.on("ready", () => {
      conn.sftp((err, sftp) => {
        if (err) {
          updateJobProgress(job.id, 0, err.message);
          conn.end();
          reject(err);
          return;
        }

        // Create remote directories
        const dirs = path.dirname(fullDest).split("/").filter(Boolean);
        let currentDir = "/";
        const mkdirRecursive = (index: number) => {
          if (index >= dirs.length) {
            doTransfer(sftp);
            return;
          }
          currentDir += (currentDir === "/" ? "" : "/") + dirs[index];
          sftp.mkdir(currentDir, (mkErr) => {
            // Ignore EEXIST errors
            mkdirRecursive(index + 1);
          });
        };

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const doTransfer = (sftpStream: any) => {
          const totalSize = job.fileSize;
          let transferred = 0;

          const readStream = fs.createReadStream(job.sourcePath);
          const writeStream = sftpStream.createWriteStream(fullDest);

          readStream.on("data", (chunk) => {
            transferred += chunk.length;
            const progress = Math.min(transferred / totalSize, 1);
            updateJobProgress(job.id, progress);
          });

          readStream.on("error", (readErr) => {
            updateJobProgress(job.id, transferred / totalSize, readErr.message);
            conn.end();
            reject(readErr);
          });

          writeStream.on("error", (writeErr: Error) => {
            updateJobProgress(
              job.id,
              transferred / totalSize,
              writeErr.message
            );
            conn.end();
            reject(writeErr);
          });

          writeStream.on("close", () => {
            updateJobProgress(job.id, 1);
            db.update(jobs)
              .set({
                destinationId: dest.id,
                destinationPath: fullDest,
                updatedAt: new Date().toISOString(),
              })
              .where(eq(jobs.id, job.id))
              .run();
            conn.end();
            resolve();
          });

          readStream.pipe(writeStream);
        };

        mkdirRecursive(0);
      });
    });

    conn.on("error", (connErr) => {
      updateJobProgress(job.id, 0, connErr.message);
      reject(connErr);
    });

    const connectConfig: Record<string, unknown> = {
      host: dest.sshHost!,
      port: dest.sshPort || 22,
      username: dest.sshUser!,
    };

    if (dest.sshKeyPath) {
      connectConfig.privateKey = fs.readFileSync(dest.sshKeyPath);
      if (dest.sshKeyPassphrase) {
        connectConfig.passphrase = dest.sshKeyPassphrase;
      }
    }

    conn.connect(connectConfig);
  });
}

/**
 * Process a single transfer
 */
async function processTransfer(jobId: number, destinationId: number) {
  activeTransfers++;

  try {
    const job = db.select().from(jobs).where(eq(jobs.id, jobId)).get();
    const dest = db
      .select()
      .from(destinations)
      .where(eq(destinations.id, destinationId))
      .get();

    if (!job || !dest) {
      throw new Error("Job or destination not found");
    }

    // Mark as transferring
    db.update(jobs)
      .set({
        status: "transferring",
        transferProgress: 0,
        transferError: null,
        updatedAt: new Date().toISOString(),
      })
      .where(eq(jobs.id, jobId))
      .run();

    const relativePath = buildRelativePath(job);

    if (dest.type === "ssh") {
      await transferSFTP(job, dest);
      await transferSubtitlesSFTP(jobId, relativePath, dest).catch((err) =>
        console.error("Subtitle SFTP error:", err)
      );
    } else {
      await transferLocal(job, dest);
      transferSubtitlesLocal(jobId, relativePath, dest.basePath);
    }
  } catch (err) {
    const message = err instanceof Error ? err.message : "Transfer failed";
    updateJobProgress(jobId, 0, message);
  }

  activeTransfers--;
  processQueue();
}

/**
 * Process next items from the queue
 */
function processQueue() {
  while (activeTransfers < MAX_CONCURRENT && transferQueue.length > 0) {
    const next = transferQueue.shift()!;
    processTransfer(next.jobId, next.destinationId);
  }
}

/**
 * Queue transfers for execution
 */
export function queueTransfers(
  jobIds: number[],
  destinationId: number
): { queued: number } {
  for (const jobId of jobIds) {
    transferQueue.push({ jobId, destinationId });
  }
  processQueue();
  return { queued: jobIds.length };
}

File: src-tauri/src/transfer.rs
use anyhow::Result;
use rusqlite::params;
use serde::Serialize;
use std::fs;
use std::io::{Read, Write};
use std::path::Path;
use std::sync::Arc;
use tauri::{AppHandle, Emitter};
use tokio::sync::Semaphore;
use tracing::{info, warn};

use crate::db;
use crate::models::{Destination, Job, TransferProgress};
use crate::naming;

const MAX_CONCURRENT: usize = 2;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct ProgressEvent {
    jobs: Vec<TransferProgress>,
}

fn update_job_progress(job_id: i64, progress: f64, error: Option<&str>) {
    let status = if error.is_some() {
        "failed"
    } else if progress >= 1.0 {
        "completed"
    } else {
        "transferring"
    };

    let db = db::conn();
    let _ = db.execute(
        "UPDATE jobs SET transfer_progress = ?1, transfer_error = ?2, status = ?3, updated_at = ?4 WHERE id = ?5",
        params![progress, error, status, db::now_iso(), job_id],
    );
}

fn build_relative_path(job: &Job) -> String {
    let group = job
        .group_id
        .and_then(db::get_group_by_id)
        .unwrap_or_else(|| {
            // Fallback for ungrouped jobs
            crate::models::Group {
                id: 0,
                status: "matched".into(),
                media_type: job.media_type.clone(),
                folder_path: String::new(),
                folder_name: String::new(),
                total_file_count: 1,
                total_file_size: job.file_size,
                parsed_title: job.parsed_title.clone(),
                parsed_year: job.parsed_year,
                tmdb_id: job.tmdb_id,
                tmdb_title: job.tmdb_title.clone(),
                tmdb_year: job.tmdb_year,
                tmdb_poster_path: job.tmdb_poster_path.clone(),
                match_confidence: job.match_confidence,
                destination_id: job.destination_id,
                created_at: job.created_at.clone(),
                updated_at: job.updated_at.clone(),
            }
        });

    let settings_map = db::get_all_settings();
    let ns = naming::NamingSettings {
        naming_preset: settings_map
            .get("naming_preset")
            .cloned()
            .unwrap_or_else(|| "jellyfin".into()),
        specials_folder_name: settings_map
            .get("specials_folder_name")
            .cloned()
            .unwrap_or_else(|| "Specials".into()),
        extras_folder_name: settings_map
            .get("extras_folder_name")
            .cloned()
            .unwrap_or_else(|| "Extras".into()),
    };

    naming::format_grouped_path(job, &group, &ns)
}

/// Local file copy with resume support
fn transfer_local(job: &Job, dest: &Destination) -> Result<()> {
    let relative = build_relative_path(job);
    let full_dest = Path::new(&dest.base_path).join(&relative);

    // Create directory structure
    if let Some(parent) = full_dest.parent() {
        fs::create_dir_all(parent)?;
    }

    let total_size = job.file_size as u64;
    let mut transferred: u64 = 0;

    // Check for partial file (resume)
    if full_dest.exists() {
        let existing_size = fs::metadata(&full_dest)?.len();
        if existing_size == total_size {
            update_job_progress(job.id, 1.0, None);
            return Ok(());
        }
        if existing_size < total_size {
            transferred = existing_size;
        }
    }

    let mut reader = fs::File::open(&job.source_path)?;
    if transferred > 0 {
        use std::io::Seek;
        reader.seek(std::io::SeekFrom::Start(transferred))?;
    }

    let mut writer = fs::OpenOptions::new()
        .create(true)
        .append(transferred > 0)
        .write(true)
        .truncate(transferred == 0)
        .open(&full_dest)?;

    let mut buf = vec![0u8; 256 * 1024]; // 256KB buffer
    loop {
        let n = reader.read(&mut buf)?;
        if n == 0 {
            break;
        }
        writer.write_all(&buf[..n])?;
        transferred += n as u64;
        let progress = (transferred as f64 / total_size as f64).min(1.0);
        update_job_progress(job.id, progress, None);
    }

    update_job_progress(job.id, 1.0, None);

    // Save destination info
    let db = db::conn();
    let _ = db.execute(
        "UPDATE jobs SET destination_id = ?1, destination_path = ?2, updated_at = ?3 WHERE id = ?4",
        params![
            dest.id,
            full_dest.to_string_lossy().to_string(),
            db::now_iso(),
            job.id,
        ],
    );

    Ok(())
}

/// SFTP transfer via ssh2
fn transfer_sftp(job: &Job, dest: &Destination) -> Result<()> {
    let relative = build_relative_path(job);
    let base = dest.base_path.replace('\\', "/");
    let rel = relative.replace('\\', "/");
    let full_dest = format!("{}/{}", base.trim_end_matches('/'), rel);

    let tcp = std::net::TcpStream::connect(format!(
        "{}:{}",
        dest.ssh_host.as_deref().unwrap_or("localhost"),
        dest.ssh_port.unwrap_or(22)
    ))?;
    tcp.set_read_timeout(Some(std::time::Duration::from_secs(30)))?;

    let mut sess = ssh2::Session::new()?;
    sess.set_tcp_stream(tcp);
    sess.handshake()?;

    // Authenticate
    if let Some(key_path) = &dest.ssh_key_path {
        let passphrase = dest.ssh_key_passphrase.as_deref();
        sess.userauth_pubkey_file(
            dest.ssh_user.as_deref().unwrap_or("root"),
            None,
            Path::new(key_path),
            passphrase,
        )?;
    } else {
        sess.userauth_agent(dest.ssh_user.as_deref().unwrap_or("root"))?;
    }

    let sftp = sess.sftp()?;

    // Create remote directories
    let dir_path = full_dest.rsplit_once('/').map(|(d, _)| d).unwrap_or("");
    let mut current = String::new();
    for part in dir_path.split('/').filter(|p| !p.is_empty()) {
        current.push('/');
        current.push_str(part);
        let _ = sftp.mkdir(Path::new(&current), 0o755);
    }

    // Transfer file
    let total_size = job.file_size as u64;
    let mut transferred: u64 = 0;

    let mut reader = fs::File::open(&job.source_path)?;
    let mut remote_file = sftp.create(Path::new(&full_dest))?;

    let mut buf = vec![0u8; 256 * 1024];
    loop {
        let n = reader.read(&mut buf)?;
        if n == 0 {
            break;
        }
        remote_file.write_all(&buf[..n])?;
        transferred += n as u64;
        let progress = (transferred as f64 / total_size as f64).min(1.0);
        update_job_progress(job.id, progress, None);
    }

    update_job_progress(job.id, 1.0, None);

    // Save destination info
    let db = db::conn();
    let _ = db.execute(
        "UPDATE jobs SET destination_id = ?1, destination_path = ?2, updated_at = ?3 WHERE id = ?4",
        params![dest.id, full_dest, db::now_iso(), job.id],
    );

    Ok(())
}

/// Process a single transfer job
async fn process_transfer(job_id: i64, destination_id: i64) {
    // Mark as transferring
    {
        let db = db::conn();
        let _ = db.execute(
            "UPDATE jobs SET status = 'transferring', transfer_progress = 0, transfer_error = NULL, updated_at = ?1 WHERE id = ?2",
            params![db::now_iso(), job_id],
        );
    }

    let job = {
        let db = db::conn();
        db.query_row(
            "SELECT * FROM jobs WHERE id = ?1",
            params![job_id],
            db::row_to_job,
        )
        .ok()
    };

    let dest = db::get_destination_by_id(destination_id);

    match (job, dest) {
        (Some(job), Some(dest)) => {
            let result = if dest.dest_type == "ssh" {
                // Run blocking SSH in a separate thread
                let j = job.clone();
                let d = dest.clone();
                tokio::task::spawn_blocking(move || transfer_sftp(&j, &d)).await
            } else {
                let j = job.clone();
                let d = dest.clone();
                tokio::task::spawn_blocking(move || transfer_local(&j, &d)).await
            };

            match result {
                Ok(Ok(())) => {
                    info!("Transfer complete for job {}", job_id);
                }
                Ok(Err(e)) => {
                    warn!("Transfer failed for job {}: {}", job_id, e);
                    update_job_progress(job_id, 0.0, Some(&e.to_string()));
                }
                Err(e) => {
                    warn!("Transfer task panic for job {}: {}", job_id, e);
                    update_job_progress(job_id, 0.0, Some(&e.to_string()));
                }
            }
        }
        _ => {
            update_job_progress(job_id, 0.0, Some("Job or destination not found"));
        }
    }
}

/// Queue and run transfers with concurrency limit
pub async fn queue_transfers(
    app: AppHandle,
    job_ids: Vec<i64>,
    destination_id: i64,
) -> Result<usize> {
    let count = job_ids.len();
    let semaphore = Arc::new(Semaphore::new(MAX_CONCURRENT));
    let app = Arc::new(app);

    // Spawn a background task that emits progress events
    let progress_app = Arc::clone(&app);
    let progress_ids = job_ids.clone();
    tokio::spawn(async move {
        loop {
            tokio::time::sleep(std::time::Duration::from_millis(200)).await;

            let progress_jobs: Vec<TransferProgress> = {
                let db = db::conn();
                let placeholders: String = progress_ids
                    .iter()
                    .map(|_| "?")
                    .collect::<Vec<_>>()
                    .join(",");
                let sql = format!(
                    "SELECT id, status, file_name, file_size, transfer_progress, transfer_error, destination_path FROM jobs WHERE id IN ({})",
                    placeholders
                );
                let mut stmt = match db.prepare(&sql) {
                    Ok(s) => s,
                    Err(_) => break,
                };
                let params: Vec<Box<dyn rusqlite::types::ToSql>> = progress_ids
                    .iter()
                    .map(|id| Box::new(*id) as Box<dyn rusqlite::types::ToSql>)
                    .collect();
                let param_refs: Vec<&dyn rusqlite::types::ToSql> =
                    params.iter().map(|p| p.as_ref()).collect();
                let result: Vec<TransferProgress> = match stmt.query_map(param_refs.as_slice(), |row| {
                    Ok(TransferProgress {
                        id: row.get("id")?,
                        status: row.get("status")?,
                        file_name: row.get("file_name")?,
                        file_size: row.get("file_size")?,
                        transfer_progress: row.get("transfer_progress")?,
                        transfer_error: row.get("transfer_error")?,
                        destination_path: row.get("destination_path")?,
                    })
                }) {
                    Ok(rows) => rows.filter_map(|r| r.ok()).collect(),
                    Err(_) => break,
                };
                result
            };

            let all_done = progress_jobs
                .iter()
                .all(|j| j.status == "completed" || j.status == "failed");

            let _ = progress_app.emit("transfer-progress", &progress_jobs);

            if all_done && !progress_jobs.is_empty() {
                break;
            }
        }
    });

    // Spawn transfer tasks with semaphore
    for job_id in job_ids {
        let sem = Arc::clone(&semaphore);
        tokio::spawn(async move {
            let _permit = sem.acquire().await.unwrap();
            process_transfer(job_id, destination_id).await;
        });
    }

    Ok(count)
}
